### PR TITLE
[GameLift][FlexMatch] Update session interface for clients to make matchmaking requests

### PR DESCRIPTION
**Pull request by junbo75**
Signed-off-by: Junbo Liang <junbo@amazon.com>

image

### CODE REVIEW

**Review by lmbr-pip — COMMENTED** [review]
LG mostly.

Need someone from Network/Framework side on this PR probably.

**Comment by junbo75** [comment]
> LG mostly.
> 
> Need someone from Network/Framework side on this PR probably.

Just added sig-network for review.

File: Gems/AWSGameLift/Code/AWSGameLiftClient/Include/Request/AWSGameLiftStartMatchmakingRequest.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        AZ_RTTI(AWSGameLiftPlayerInformation, "{B62C118E-C55D-4903-8ECB-E58E8CA613C4}");
+        static void Reflect(AZ::ReflectContext* context);
+
+        AWSGameLiftPlayerInformation() = default;
+        virtual ~AWSGameLiftPlayerInformation() = default;
+
+        // A set of values, expressed in milliseconds, that indicates
+        // the amount of latency that a player experiences when connected to AWS Regions
+        AZStd::unordered_map<AZStd::string, int> m_latencyInMs;
```
[the remark this answers is not in the document]
Comment (junbo75) [reply Gems/AWSGameLift/Code/AWSGameLiftClient/Include/Request/AWSGameLiftStartMatchmakingRequest.h:29]: Sure. Will do.

File: Gems/AWSGameLift/Code/AWSGameLiftClient/Include/Request/AWSGameLiftStartMatchmakingRequest.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        // A collection of key:value pairs containing player information for use in matchmaking
+        // Player attribute keys must match the playerAttributes used in a matchmaking rule set
+        // Example: {"skill": "{\"N\": \"23\"}", "gameMode": "{\"S\": \"deathmatch\"}"}
+        AZStd::unordered_map<AZStd::string, AZStd::string> m_playerAttributes;
```
Comment (onecent1101) [inline Gems/AWSGameLift/Code/AWSGameLiftClient/Include/Request/AWSGameLiftStartMatchmakingRequest.h:33]: So matchmaking request takes json format?

**Review by onecent1101 — COMMENTED** [review]
As discussed offline, don't forget to add a notification for accept match

**Comment by AMZN-puvvadar** [comment]
From a cursory look, I don't see issues from the networking perspective. This change largely looks to implement a service provider for ISession interfaces.

**Comment by AMZN-puvvadar** [comment]
It seems very odd that there is a file named ISessionRequests.cpp. It seems that it's largely for context exposure purposes? What's the intention and is there better naming/structure here?

**Comment by junbo75** [comment]
> It seems very odd that there is a file named ISessionRequests.cpp. It seems that it's largely for context exposure purposes? What's the intention and is there better naming/structure here?

**Comment by junbo75** [comment]
> It seems very odd that there is a file named ISessionRequests.cpp. It seems that it's largely for context exposure purposes? What's the intention and is there better naming/structure here?

Talked offline. Will add sig-core as reviewers to check the suggested style here.

File: Code/Framework/AzFramework/AzFramework/Session/ISessionRequests.cpp [hunk 3]
```diff
@@ -13,6 +13,31 @@
 
 namespace AzFramework
 {
+    void AcceptMatchRequest::Reflect(AZ::ReflectContext* context)
```
[the remark this answers is not in the document]
Comment (junbo75) [reply Code/Framework/AzFramework/AzFramework/Session/ISessionRequests.cpp:16]: Talked offline. Will move the matchmaking related requests/notifications to separate files.

File: Gems/AWSGameLift/Code/AWSGameLiftClient/Include/Request/AWSGameLiftStartMatchmakingRequest.h (hunk 2 above)
Comment (junbo75) [reply Gems/AWSGameLift/Code/AWSGameLiftClient/Include/Request/AWSGameLiftStartMatchmakingRequest.h:33]: GameLift defines its own data structure we need to convert to before sending the request.

**Comment by junbo75** [comment]
Add sig-core to check the best practice for defining interface in AZ Framework and implement base type reflection according to the discussion with @AMZN-puvvadar.

**Review by onecent1101 — APPROVED** [review]
Confirm with sig-core on styling, then im good with it.

**Review by onecent1101 — APPROVED** [review]

File: Gems/AWSGameLift/Code/AWSGameLiftClient/Source/AWSGameLiftClientManager.cpp [hunk 4]
```diff
@@ -349,6 +359,28 @@ namespace AWSGameLift
         return response;
     }
 
+    AZStd::string AWSGameLiftClientManager::StartMatchmaking(const AzFramework::StartMatchmakingRequest& startMatchmakingRequest)
+    {
+        AZ_UNUSED(startMatchmakingRequest);
+
+        return "";
```
Comment (lumberyard-employee-dm) [inline Gems/AWSGameLift/Code/AWSGameLiftClient/Source/AWSGameLiftClientManager.cpp:366]: FYI returning `AZStd::string{}` is more efficient.
`""` needs to at least have `strlen` invoked on it to figure out it is zero bytes in lenght.
A default constructed string doesn't have to deal with that

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by AMZN-puvvadar — APPROVED** [review]

**Pull request merged**